### ISSUE TITLE
rktio: IPv6 unreachability can cause very slow TCP connection times

### ISSUE THREAD

**Issue by lexi-lambda** [racket/racket#3941]
On my machine, connecting to `pkgs.racket-lang.org` via TCP can easily take a full 6 seconds:

```
$ racket -e '(require racket/tcp) (time (tcp-connect "pkgs.racket-lang.org" 80))'
cpu time: 1 real time: 6090 gc time: 0
```

After a fair bit of investigation, I have tracked down the root cause. When rktio resolves `pkgs.racket-lang.org` using `getaddrinfo`, it receives 4 results: two IPv6 addresses and two IPv4 addresses. Both IPv6 addresses fail when passed to `connect` with a “no route to host” error, but the IPv4 address eventually succeeds. Unfortunately, it can take a full 3 seconds before each `connect` attempt fails, leading to unbearably slow TCP connection times. (This is particularly bad in the case of `pkgs.racket-lang.org`, since it is hammered by `raco pkg install` to look up package information.)

The IETF recommended solution for this is known as [Happy Eyeballs](https://en.wikipedia.org/wiki/Happy_Eyeballs) (for reasons that elude me), which involves trying to open both an IPv4 and an IPv6 connection concurrently. Happy Eyeballs is implemented by tools like `curl` to speed up connections in situations like mine, and it would be nice if `rktio` were to do so as well.

**Comment by samth** [comment]
Do you know why your machine is unable to use the v6 addresses successfully?

**Comment by lexi-lambda** [comment]
It doesn’t seem to be a function of my machine *per se*, but rather a function of my ISP, which is obviously out of my control.

**Comment by samth** [comment]
Right, obviously we can improve this experience in Racket. I was just wondering what the ipv6 failure was (sigh, in 2021).

**Comment by lexi-lambda** [comment]
Yes, that’s reasonable. I’m afraid I don’t have much insight into why my ISP doesn’t have a working IPv6 configuration, but as far as I can tell, the failure is both total and immediate.

**Comment by lexi-lambda** [comment]
In the meantime, as a workaround, I’ve figured out how to disable IPv6 on a particular NetworkManager-managed connection (for those using both Linux and NetworkManager). Specifically,

```
$ nmcli connection modify $connection_name ipv6.method disabled
$ nmcli connection up $connection_name
```

seems to do the trick.
